File: NoteBase_csharp/Storage/MetaJson.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using NoteBase.Core;

namespace NoteBase.Storage
{
    /// <summary>
    /// meta.json 専用の JSON シリアライザ／デシリアライザ。
    /// 標準ライブラリのみで実装するため、自前で読み書きを行う。
    /// 出力はインデント付き、フィールド順固定、null 値は省略する。
    /// </summary>
    public static class MetaJson
    {
        // ============================================================
        // Serialize
        // ============================================================

        public static string Serialize(NoteMeta m)
        {
            var lines = new List<string>();
            lines.Add(IntField("meta_version", m.MetaVersion));
            lines.Add(StringField("id", m.Id));
            lines.Add(StringField("title", m.Title));
            lines.Add(StringField("type", m.Type.ToWireString()));
            if (m.Status.HasValue)
                lines.Add(StringField("status", m.Status.Value.ToWireString()));
            lines.Add(StringArrayField("tags", m.Tags ?? new List<string>()));
            if (!string.IsNullOrEmpty(m.Project))
                lines.Add(StringField("project", m.Project));
            if (m.Due.HasValue)
                lines.Add(StringField("due", FormatDate(m.Due.Value)));
            if (m.Schedule != null)
                lines.Add(ScheduleField("schedule", m.Schedule));
            if (!string.IsNullOrEmpty(m.InstanceOf))
                lines.Add(StringField("instance_of", m.InstanceOf));
            lines.Add(StringField("created", FormatDateTime(m.Created)));
            lines.Add(StringField("updated", FormatDateTime(m.Updated)));

            return "{\n" + string.Join(",\n", lines) + "\n}\n";
        }

        private static string IntField(string name, int value)
        {
            return "  \"" + name + "\": " + value.ToString(CultureInfo.InvariantCulture);
        }

        private static string StringField(string name, string value)
        {
            return "  \"" + name + "\": " + EscapeString(value ?? "");
        }

        private static string StringArrayField(string name, List<string> values)
        {
            var parts = new List<string>();
            foreach (var v in values) parts.Add(EscapeString(v ?? ""));
            return "  \"" + name + "\": [" + string.Join(", ", parts) + "]";
        }

        private static string ScheduleField(string name, Schedule s)
        {
            var sb = new StringBuilder();
            sb.Append("  \"").Append(name).Append("\": {\n");
            var inner = new List<string>();
            inner.Add("    \"frequency\": " + EscapeString(s.Frequency.ToWireString()));
            if (s.Days != null && s.Days.Count > 0)
            {
                var parts = new List<string>();
                foreach (var d in s.Days) parts.Add(EscapeString(d ?? ""));
                inner.Add("    \"days\": [" + string.Join(", ", parts) + "]");
            }
            if (s.DayOfMonth.HasValue)
                inner.Add("    \"day_of_month\": " + s.DayOfMonth.Value.ToString(CultureInfo.InvariantCulture));
            inner.Add("    \"enabled\": " + (s.Enabled ? "true" : "false"));
            sb.Append(string.Join(",\n", inner));
            sb.Append("\n  }");
            return sb.ToString();
        }

        private static string EscapeString(string s)
        {
            var sb = new StringBuilder();
            sb.Append('"');
            foreach (var c in s)
            {
                switch (c)
                {
                    case '"': sb.Append("\\\""); break;
                    case '\\': sb.Append("\\\\"); break;
                    case '\b': sb.Append("\\b"); break;
                    case '\f': sb.Append("\\f"); break;
                    case '\n': sb.Append("\\n"); break;
                    case '\r': sb.Append("\\r"); break;
                    case '\t': sb.Append("\\t"); break;
                    default:
                        if (c < 0x20)
                            sb.Append("\\u").Append(((int)c).ToString("X4"));
                        else
                            sb.Append(c);
                        break;
                }
            }
            sb.Append('"');
            return sb.ToString();
        }

        private static string FormatDateTime(DateTime dt)
        {
            return dt.ToString("yyyy-MM-ddTHH:mm:ss", CultureInfo.InvariantCulture);
        }

        private static string FormatDate(DateTime dt)
        {
            return dt.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
        }

        // ============================================================
        // Deserialize
        // ============================================================

        public static NoteMeta Deserialize(string json)
        {
            var parser = new Parser(json);
            var dict = parser.ParseObject();
            return MetaFromDict(dict);
        }

        private static NoteMeta MetaFromDict(Dictionary<string, object> d)
        {
            var m = new NoteMeta();

            object v;
            if (d.TryGetValue("meta_version", out v)) m.MetaVersion = ToInt(v);
            if (d.TryGetValue("id", out v)) m.Id = (string)v;
            if (d.TryGetValue("title", out v)) m.Title = (string)v;
            if (d.TryGetValue("type", out v)) m.Type = NoteTypeExtensions.ParseType((string)v);
            if (d.TryGetValue("status", out v))
                m.Status = (v == null) ? null : NoteStatusExtensions.ParseStatus((string)v);
            if (d.TryGetValue("tags", out v))
            {
                m.Tags = new List<string>();
                var lo = v as List<object>;
                if (lo != null)
                    foreach (var o in lo) m.Tags.Add(o == null ? "" : (string)o);
            }
            if (d.TryGetValue("project", out v)) m.Project = (string)v;
            if (d.TryGetValue("due", out v))
                m.Due = (v == null) ? (DateTime?)null : ParseDate((string)v);
            if (d.TryGetValue("schedule", out v) && v is Dictionary<string, object>)
                m.Schedule = ScheduleFromDict((Dictionary<string, object>)v);
            if (d.TryGetValue("instance_of", out v)) m.InstanceOf = (string)v;
            if (d.TryGetValue("created", out v) && v is string)
                m.Created = ParseDateTime((string)v);
            if (d.TryGetValue("updated", out v) && v is string)
                m.Updated = ParseDateTime((string)v);

            return m;
        }

        private static Schedule ScheduleFromDict(Dictionary<string, object> d)
        {
            var s = new Schedule();
            object v;
            if (d.TryGetValue("frequency", out v))
                s.Frequency = ScheduleFrequencyExtensions.ParseFrequency((string)v);
            if (d.TryGetValue("days", out v))
            {
                s.Days = new List<string>();
                var lo = v as List<object>;
                if (lo != null)
                    foreach (var o in lo) s.Days.Add((string)o);
            }
            if (d.TryGetValue("day_of_month", out v))
                s.DayOfMonth = (v == null) ? (int?)null : ToInt(v);
            if (d.TryGetValue("enabled", out v))
                s.Enabled = (v is bool) ? (bool)v : true;
            return s;
        }

        private static int ToInt(object v)
        {
            if (v is long) return (int)(long)v;
            if (v is int) return (int)v;
            if (v is double) return (int)(double)v;
            return 0;
        }

        private static DateTime ParseDateTime(string s)
        {
            DateTime r;
            if (DateTime.TryParseExact(s, "yyyy-MM-ddTHH:mm:ss",
                CultureInfo.InvariantCulture, DateTimeStyles.None, out r))
                return r;
            if (DateTime.TryParse(s, CultureInfo.InvariantCulture, DateTimeStyles.None, out r))
                return r;
            return DateTime.MinValue;
        }

        private static DateTime ParseDate(string s)
        {
            DateTime r;
            if (DateTime.TryParseExact(s, "yyyy-MM-dd",
                CultureInfo.InvariantCulture, DateTimeStyles.None, out r))
                return r;
            if (DateTime.TryParse(s, CultureInfo.InvariantCulture, DateTimeStyles.None, out r))
                return r;
            return DateTime.MinValue;
        }

        // ============================================================
        // 内部 JSON パーサー
        // ============================================================

        private class Parser
        {
            private readonly string _src;
            private int _pos;

            public Parser(string src)
            {
                _src = src ?? "";
                _pos = 0;
            }

            public Dictionary<string, object> ParseObject()
            {
                SkipWs();
                Expect('{');
                var dict = new Dictionary<string, object>();
                SkipWs();
                if (Peek() == '}') { _pos++; return dict; }
                while (true)
                {
                    SkipWs();
                    var key = ParseString();
                    SkipWs();
                    Expect(':');
                    var value = ParseValue();
                    dict[key] = value;
                    SkipWs();
                    char nc = Peek();
                    if (nc == ',') { _pos++; continue; }
                    if (nc == '}') { _pos++; return dict; }
                    throw new FormatException("Expected ',' or '}' at pos " + _pos);
                }
            }

            private object ParseValue()
            {
                SkipWs();
                if (_pos >= _src.Length)
                    throw new FormatException("Unexpected end of JSON");
                char c = _src[_pos];
                if (c == '{') return ParseObject();
                if (c == '[') return ParseArray();
                if (c == '"') return ParseString();
                if (c == 't' || c == 'f') return ParseBool();
                if (c == 'n') { ParseNull(); return null; }
                if (c == '-' || (c >= '0' && c <= '9')) return ParseNumber();
                throw new FormatException("Unexpected character at pos " + _pos + ": " + c);
            }

            private List<object> ParseArray()
            {
                Expect('[');
                var list = new List<object>();
                SkipWs();
                if (Peek() == ']') { _pos++; return list; }
                while (true)
                {
                    list.Add(ParseValue());
                    SkipWs();
                    char nc = Peek();
                    if (nc == ',') { _pos++; continue; }
                    if (nc == ']') { _pos++; return list; }
                    throw new FormatException("Expected ',' or ']' at pos " + _pos);
                }
            }

            private string ParseString()
            {
                Expect('"');
                var sb = new StringBuilder();
                while (_pos < _src.Length)
                {
                    char c = _src[_pos++];
                    if (c == '"') return sb.ToString();
                    if (c == '\\')
                    {
                        if (_pos >= _src.Length)
                            throw new FormatException("Unterminated escape");
                        char esc = _src[_pos++];
                        switch (esc)
                        {
                            case '"': sb.Append('"'); break;
                            case '\\': sb.Append('\\'); break;
                            case '/': sb.Append('/'); break;
                            case 'b': sb.Append('\b'); break;
                            case 'f': sb.Append('\f'); break;
                            case 'n': sb.Append('\n'); break;
                            case 'r': sb.Append('\r'); break;
                            case 't': sb.Append('\t'); break;
                            case 'u':
                                if (_pos + 4 > _src.Length)
                                    throw new FormatException("Bad \\u escape");
                                var hex = _src.Substring(_pos, 4);
                                _pos += 4;
                                sb.Append((char)int.Parse(hex,
                                    NumberStyles.HexNumber, CultureInfo.InvariantCulture));
                                break;
                            default:
                                throw new FormatException("Unknown escape: \\" + esc);
                        }
                    }
                    else
                    {
                        sb.Append(c);
                    }
                }
                throw new FormatException("Unterminated string");
            }

            private object ParseNumber()
            {
                int start = _pos;
                if (_src[_pos] == '-') _pos++;
                while (_pos < _src.Length)
                {
                    char c = _src[_pos];
                    if (char.IsDigit(c) || c == '.' || c == 'e' || c == 'E' || c == '+' || c == '-')
                        _pos++;
                    else break;
                }
                var s = _src.Substring(start, _pos - start);
                if (s.IndexOf('.') < 0 && s.IndexOf('e') < 0 && s.IndexOf('E') < 0)
                {
                    long lv;
                    if (long.TryParse(s, NumberStyles.Integer, CultureInfo.InvariantCulture, out lv))
                        return lv;
                }
                double dv;
                if (double.TryParse(s, NumberStyles.Float, CultureInfo.InvariantCulture, out dv))
                    return dv;
                throw new FormatException("Invalid number: " + s);
            }

            private bool ParseBool()
            {
                if (Match("true")) return true;
                if (Match("false")) return false;
                throw new FormatException("Expected bool at pos " + _pos);
            }

            private void ParseNull()
            {
                if (Match("null")) return;
                throw new FormatException("Expected null at pos " + _pos);
            }

            private bool Match(string keyword)
            {
                if (_pos + keyword.Length > _src.Length) return false;
                if (_src.Substring(_pos, keyword.Length) != keyword) return false;
                _pos += keyword.Length;
                return true;
            }

            private void Expect(char c)
            {
                SkipWs();
                if (_pos >= _src.Length || _src[_pos] != c)
                    throw new FormatException("Expected '" + c + "' at pos " + _pos);
                _pos++;
            }

            private char Peek()
            {
                SkipWs();
                if (_pos >= _src.Length)
                    throw new FormatException("Unexpected end of JSON");
                return _src[_pos];
            }

            private void SkipWs()
            {
                while (_pos < _src.Length)
                {
                    char c = _src[_pos];
                    if (c == ' ' || c == '\t' || c == '\n' || c == '\r') _pos++;
                    else break;
                }
            }
        }
    }
}

File: notebase/storage/meta_json.py
from __future__ import annotations

import json
from datetime import datetime

from ..core.note_meta import NoteMeta
from ..core.note_status import NoteStatus
from ..core.note_type import NoteType
from ..core.schedule import Schedule, ScheduleFrequency


def serialize(m: NoteMeta) -> str:
    """NoteMeta を C# 版とバイト一致する JSON 文字列に変換する。

    - フィールド順固定
    - インデント 2 スペース
    - 空でない/値ありフィールドのみ出力 (status/project/due/schedule/instance_of は省略可)
    - tags は常に出力 (空でも [] を出す)
    - 末尾に改行 1 つ
    """
    lines: list[str] = []
    lines.append(_int_field("meta_version", m.meta_version))
    lines.append(_string_field("id", m.id or ""))
    lines.append(_string_field("title", m.title or ""))
    lines.append(_string_field("type", m.type.to_wire()))
    if m.status is not None:
        lines.append(_string_field("status", m.status.to_wire()))
    lines.append(_string_array_field("tags", m.tags or []))
    if m.project:
        lines.append(_string_field("project", m.project))
    if m.due is not None:
        lines.append(_string_field("due", _format_date(m.due)))
    if m.schedule is not None:
        lines.append(_schedule_field("schedule", m.schedule))
    if m.instance_of:
        lines.append(_string_field("instance_of", m.instance_of))
    if m.estimated_minutes is not None:
        lines.append(_int_field("estimated_minutes", m.estimated_minutes))
    if m.actual_minutes is not None:
        lines.append(_int_field("actual_minutes", m.actual_minutes))
    lines.append(_string_field("created", _format_datetime(m.created)))
    lines.append(_string_field("updated", _format_datetime(m.updated)))

    return "{\n" + ",\n".join(lines) + "\n}\n"


def deserialize(text: str) -> NoteMeta:
    """JSON テキストから NoteMeta を復元する。"""
    d = json.loads(text)
    return _meta_from_dict(d)


def _meta_from_dict(d: dict) -> NoteMeta:
    m = NoteMeta()
    if "meta_version" in d and d["meta_version"] is not None:
        m.meta_version = int(d["meta_version"])
    if "id" in d:
        m.id = d["id"] or ""
    if "title" in d:
        m.title = d["title"] or ""
    if "type" in d:
        m.type = NoteType.parse(d["type"])
    if "status" in d:
        m.status = NoteStatus.parse(d["status"])
    if "tags" in d and isinstance(d["tags"], list):
        m.tags = [v if v is not None else "" for v in d["tags"]]
    if "project" in d:
        m.project = d["project"]
    if "due" in d and d["due"] is not None:
        m.due = _parse_date(d["due"])
    if "schedule" in d and isinstance(d["schedule"], dict):
        m.schedule = _schedule_from_dict(d["schedule"])
    if "instance_of" in d:
        m.instance_of = d["instance_of"]
    if "estimated_minutes" in d and d["estimated_minutes"] is not None:
        try:
            m.estimated_minutes = int(d["estimated_minutes"])
        except (TypeError, ValueError):
            m.estimated_minutes = None
    if "actual_minutes" in d and d["actual_minutes"] is not None:
        try:
            m.actual_minutes = int(d["actual_minutes"])
        except (TypeError, ValueError):
            m.actual_minutes = None
    if "created" in d and isinstance(d["created"], str):
        m.created = _parse_datetime(d["created"])
    if "updated" in d and isinstance(d["updated"], str):
        m.updated = _parse_datetime(d["updated"])
    return m


def _schedule_from_dict(d: dict) -> Schedule:
    s = Schedule()
    if "frequency" in d:
        s.frequency = ScheduleFrequency.parse(d["frequency"])
    if "days" in d and isinstance(d["days"], list):
        s.days = [v for v in d["days"] if isinstance(v, str)]
    if "day_of_month" in d and d["day_of_month"] is not None:
        s.day_of_month = int(d["day_of_month"])
    if "enabled" in d:
        s.enabled = bool(d["enabled"]) if isinstance(d["enabled"], bool) else True
    return s


def _int_field(name: str, value: int) -> str:
    return f'  "{name}": {int(value)}'


def _string_field(name: str, value: str) -> str:
    return f'  "{name}": {_escape_string(value or "")}'


def _string_array_field(name: str, values: list[str]) -> str:
    parts = [_escape_string(v or "") for v in values]
    return f'  "{name}": [' + ", ".join(parts) + "]"


def _schedule_field(name: str, s: Schedule) -> str:
    inner: list[str] = []
    inner.append(f'    "frequency": {_escape_string(s.frequency.to_wire())}')
    if s.days:
        parts = [_escape_string(d or "") for d in s.days]
        inner.append('    "days": [' + ", ".join(parts) + "]")
    if s.day_of_month is not None:
        inner.append(f'    "day_of_month": {int(s.day_of_month)}')
    inner.append(f'    "enabled": {"true" if s.enabled else "false"}')
    body = ",\n".join(inner)
    return f'  "{name}": {{\n{body}\n  }}'


def _escape_string(s: str) -> str:
    out: list[str] = ['"']
    for ch in s:
        c = ord(ch)
        if ch == '"':
            out.append('\\"')
        elif ch == "\\":
            out.append("\\\\")
        elif ch == "\b":
            out.append("\\b")
        elif ch == "\f":
            out.append("\\f")
        elif ch == "\n":
            out.append("\\n")
        elif ch == "\r":
            out.append("\\r")
        elif ch == "\t":
            out.append("\\t")
        elif c < 0x20:
            out.append(f"\\u{c:04X}")
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


def _format_datetime(dt: datetime) -> str:
    return dt.strftime("%Y-%m-%dT%H:%M:%S")


def _format_date(dt: datetime) -> str:
    return dt.strftime("%Y-%m-%d")


def _parse_datetime(s: str) -> datetime:
    try:
        return datetime.strptime(s, "%Y-%m-%dT%H:%M:%S")
    except ValueError:
        try:
            return datetime.fromisoformat(s)
        except ValueError:
            return datetime(1, 1, 1)


def _parse_date(s: str) -> datetime:
    try:
        return datetime.strptime(s, "%Y-%m-%d")
    except ValueError:
        try:
            return datetime.fromisoformat(s)
        except ValueError:
            return datetime(1, 1, 1)
